File: backend/app/api/spots.py
"""Scenic spots API endpoints."""
import logging

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.models.tourist import ScenicSpot
from app.services.scenic_fallback import get_demo_spot, list_demo_spots

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/api/spots", tags=["spots"])


class SpotOut(BaseModel):
    id: str
    name: str
    category: str
    tags: list[str] | None
    overview: str
    qr_code: str | None
    thumbnail: str | None
    duration: str | None
    display_x: float | None
    display_y: float | None
    latitude: float | None
    longitude: float | None
    qa_json: list[dict] | None

    class Config:
        from_attributes = True


class SpotDetail(SpotOut):
    detail: str
    related_spots: list[str] | None
    detail_images: list[str] | None
    story_acts: list[dict] | None
    qa_json: list[dict] | None


@router.get("", response_model=list[SpotOut])
async def list_spots(
    category: str | None = Query(None),
    db: AsyncSession = Depends(get_db),
):
    """List all scenic spots, optionally filtered by category."""
    try:
        stmt = select(ScenicSpot).where(ScenicSpot.is_active == True)
        if category:
            stmt = stmt.where(ScenicSpot.category == category)
        stmt = stmt.order_by(ScenicSpot.name)
        result = await db.execute(stmt)
        spots = result.scalars().all()
        return [SpotOut.model_validate(s) for s in spots]
    except Exception as exc:
        logger.warning("Failed to query spots: %s, returning demo data", exc)
        return [SpotOut.model_validate(s) for s in list_demo_spots(category)]


@router.get("/{spot_id}", response_model=SpotDetail)
async def get_spot(
    spot_id: str,
    db: AsyncSession = Depends(get_db),
):
    """Get a scenic spot by ID."""
    try:
        stmt = select(ScenicSpot).where(ScenicSpot.id == spot_id)
        result = await db.execute(stmt)
        spot = result.scalar_one_or_none()
        if spot:
            return SpotDetail.model_validate(spot)
    except Exception as exc:
        logger.warning("Failed to query spot %s: %s, trying demo data", spot_id, exc)

    demo_spot = get_demo_spot(spot_id)
    if not demo_spot:
        raise HTTPException(status_code=404, detail="景点未找到")
    return SpotDetail.model_validate(demo_spot)


@router.get("/{spot_id}/card")
async def get_spot_card(
    spot_id: str,
    db: AsyncSession = Depends(get_db),
):
    """Return structured card data for a scenic spot (for theme cards)."""
    stmt = select(ScenicSpot).where(ScenicSpot.id == spot_id)
    result = await db.execute(stmt)
    spot = result.scalar_one_or_none()
    if not spot:
        raise HTTPException(status_code=404, detail="景点未找到")

    return {
        "type": "spot_info",
        "spot_id": spot.id,
        "name": spot.name,
        "category": spot.category,
        "overview": spot.overview,
        "detail": spot.detail,
        "tags": spot.tags or [],
        "latitude": spot.latitude,
        "longitude": spot.longitude,
        "ticket_info": spot.ticket_info,
        "open_time": spot.open_time,
        "must_see": spot.must_see,
        "best_time": spot.best_time,
        "narration": spot.narration,
        "thumbnail": spot.thumbnail,
        "duration": spot.duration,
    }

File: frontend/src/api/spots.ts
import { get, getCached } from './request';

export interface Spot {
  id: string;
  name: string;
  category: string;
  tags: string[] | null;
  overview: string;
  qr_code: string | null;
  thumbnail: string | null;
  duration: string | null;
  display_x: number | null;
  display_y: number | null;
  latitude: number | null;
  longitude: number | null;
  qa_json: { q: string; a: string }[] | null;
  story_acts: StoryAct[] | null;
}

export interface StoryAct {
  id: string;
  title: string;
  emotion: string;
  prompt_hint: string;
  act_image: string;
}

export interface SpotDetail extends Spot {
  detail: string;
  related_spots: string[] | null;
  detail_images: string[] | null;
  story_acts: StoryAct[] | null;
  qa_json: { q: string; a: string }[] | null;
}

export interface TourRoute {
  id: string;
  name: string;
  route_type: string;
  duration: string;
  description: string;
  gradient: string | null;
  cover_image: string | null;
  color: string | null;
  brush_image: string | null;
  opening_text: string | null;
  closing_text: string | null;
  spot_order: string[];
  spot_details: Record<string, unknown> | null;
  is_active: boolean;
}

export const listSpots = async (category?: string): Promise<Spot[]> => {
  const resp = await getCached<Spot[]>('/spots', category ? { category } : undefined, 60000);
  return resp.data;
};

export const getSpotById = async (id: string): Promise<SpotDetail> => {
  const resp = await get<SpotDetail>(`/spots/${id}`);
  return resp.data;
};

export const listRoutes = async (): Promise<TourRoute[]> => {
  const resp = await get<TourRoute[]>('/routes');
  return resp.data;
};

export const getRouteById = async (id: string): Promise<TourRoute> => {
  const resp = await get<TourRoute>(`/routes/${id}`);
  return resp.data;
};
